### PR TITLE
cli: list entry points command

### PR DESCRIPTION
**Pull request by lnielsen**
* Adds command to list known entry points in an a given Python
  environment.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

### CODE REVIEW

**Comment by lnielsen** [comment]
screen shot 2015-12-02 at 10 08 28

**Comment by jirikuncar** [comment]
This is very useful!

--
PS: I guess you are still working on tests :wink:

**Comment by lnielsen** [comment]
Yep :-) Unfortunately tests didn't pass in first go :-)

**Comment by lnielsen** [comment]
@jirikuncar Ready - note, there's two extra smaller commits added compared to first version.

File: invenio_base/app.py [hunk 1]
```diff
@@ -181,10 +181,19 @@ def _loader(init_func, entry_points=None, modules=None):
...
+                    app.logger.error(
+                        "Failed to initialize entry point: {0}".format(ep))
+                    raise
     if modules:
         for m in modules:
-            init_func(m)
+            try:
+                init_func(m)
+            except Exception:
+                app.logger.error("Failed to initialize module: {0}".format(m))
```
Comment (jirikuncar) [inline invenio_base/app.py]: error vs exception?

File: invenio_base/app.py [hunk 2]
```diff
@@ -181,10 +181,19 @@ def _loader(init_func, entry_points=None, modules=None):
     if entry_points:
         for entry_point in entry_points:
             for ep in pkg_resources.iter_entry_points(entry_point):
-                init_func(ep.load())
+                try:
+                    init_func(ep.load())
+                except Exception:
+                    app.logger.error(
```
Comment (jirikuncar) [inline invenio_base/app.py]: error vs. exception?

File: tests/test_app.py [hunk 3]
```diff
@@ -151,6 +158,28 @@ def test_loader():
...
+    app = Flask(__name__)
+    handler = MockLoggingHandler()
+    app.logger.addHandler(handler)
+
+    def _raise_func():
+        raise Exception()
+
+    assert len(handler.messages['error']) == 0
+    pytest.raises(
+        Exception, _loader, app, lambda x: x(),
```
Comment (jirikuncar) [inline tests/test_app.py]: `lambda x: x()` should be `x`, or am I missing something?

File: invenio_base/app.py (hunk 2 above)
Comment (lnielsen) [inline invenio_base/app.py]: Exception is re-raised, so some logger higher up the hierarchy will catch it and cause double exception to be displayed if I used app.logger.exception.

The problem above code solves is if e.g. entry point loading fails, then you wouldn't see which modules was causing the problem and thus be blind as to where the problem was.

Comment (jirikuncar) [inline invenio_base/app.py]: Ok

**Pull request merged**